src/main/java/com/google/sps/servlets/TasksServlet.java

```diff
@@ -61,13 +62,72 @@ public TasksServlet(
+  /**
+   * Returns Tasks from the user's Tasks account
+   *
+   * @param request Http request from client. Should contain idToken and accessToken
+   * @param response 403 if user is not authenticated, list of Tasks otherwise
+   * @param googleCredential a valid google credential object (already verified)
+   * @throws IOException if an issue arises while processing the request
+   */
+  @Override
+  public void doGet(
+      HttpServletRequest request, HttpServletResponse response, Credential googleCredential)
+      throws IOException {
+    assert googleCredential != null
+        : "Null credentials (i.e. unauthenticated requests) should already be handled";
+
+    // Get tasks from Google Tasks
+    TasksClient tasksClient = tasksClientFactory.getTasksClient(googleCredential);
     List<TaskList> taskLists = tasksClient.listTaskLists();
-    List<Task> tasks = new ArrayList<>();
-    for (TaskList taskList : taskLists) {
-      tasks.addAll(tasksClient.listTasks(taskList));
+    List<Task> tasks = getTasks(tasksClient);
+
+    // Initialize Tasks Response
+    List<String> taskListTitles = getTaskListTitles(taskLists);
+    int tasksToComplete = getTasksToComplete(tasks);
+    int tasksDueToday = getTasksDueToday(tasks);
+    int tasksCompletedToday = getTasksCompletedToday(tasks);
+    int tasksOverdue = getTasksOverdue(tasks);
+    TasksResponse tasksResponse =
+        new TasksResponse(
+            taskListTitles, tasksToComplete, tasksDueToday, tasksCompletedToday, tasksOverdue);
+
+    // Convert tasks to JSON and print to response
+    JsonUtility.sendJson(response, tasksResponse);
+  }
+
+  /**
+   * Add a new task object to a tasklist
+   *
+   * @param request HTTP request from client. Must contain a taskListId. Body must contain a task
+   *     entity (https://developers.google.com/tasks/v1/reference/tasks)
+   * @param response Http response to be sent to client
+   * @param googleCredential valid, verified google credential object
+   * @throws IOException if an issue occurs with the tasksService
+   */
+  @Override
+  public void doPost(
+      HttpServletRequest request, HttpServletResponse response, Credential googleCredential)
+      throws IOException {
+    assert googleCredential != null
+        : "Null credentials (i.e. unauthenticated requests) should already be handled";
+    TasksClient tasksClient = tasksClientFactory.getTasksClient(googleCredential);
+
+    String taskListId = request.getParameter("taskListId");
+
+    if (taskListId == null || taskListId.equals("")) {
+      response.sendError(400, "taskListId must be present in request");
+      return;
+    }
+
+    Gson gson = new Gson();
+    Task taskToPost = gson.fromJson(request.getReader(), Task.class);
+
+    try {
+      JsonUtility.sendJson(response, tasksClient.postTask(taskListId, taskToPost));
+    } catch (IOException e) {
+      response.sendError(400, "taskListId is invalid");
```

Comment:
Yeah good catch - so an invalid `taskListId` can cause an IOException here (Google does not have this documented - I tested this manually), however there are also tons of other reasons this could happen. 

I changed this to throw the error with a different message for the client, that indicates the `taskListId` could be the cause, but not the only cause. Let me know what you think!

Reply:
It's good, thanks for improving this.
RESOLVED